File: scher-core/src/counted.rs
//! counted.rs — the ONE implementation of the counted fold.
//!
//! MUSLIN — seams showing on purpose.
//!
//! WHY THIS FILE EXISTS
//! --------------------
//! `scher-pages/src/counted.ts` folds typed key/value over an event chain in
//! TypeScript. `scher-state/src/lib.rs` rasterizes typed key/value over a
//! succession chain in Rust. Those are the same operation, written twice, and
//! two hand-kept twins drift — which is the exact problem `scher-core-wasm`
//! was built to end ("one compiled kernel instead of two hand-kept twins",
//! Hallie's ruling 2026-07-21).
//!
//! So the fold lives HERE, compiles to wasm, and the TS becomes a caller
//! rather than a second implementation. The conformance corpus
//! (`conformance/counted.json`) is neutral ground both replay against, so
//! agreement is PROVED rather than assumed.
//!
//! THE LAW THIS ENCODES
//! --------------------
//!   * a value is FOLDED from delta-beats, never stored. There is no field to
//!     drift from its history because there is no field.
//!   * occlusion re-folds: undo is not a stack, it is a beat that stops
//!     counting.
//!   * as-of is free: the past is a read, not a snapshot anyone kept.
//!   * over-cap is REFUSED, never clamped — a clamped write is a lie the
//!     history cannot show. (Refusal lives at the write, which is caller-side;
//!     this module reads, and reports the cap so a caller can refuse.)
//!   * section membership is read from the newest live placement, never a
//!     field on the key.
//!
//! WHAT THE COMPILED KERNEL ACTUALLY BUYS (Hallie, 2026-07-30)
//! ----------------------------------------------------------
//! Not just "one implementation instead of two." The bigger prize:
//!
//!   "using much simpler and more optimal storage and code and folding it into
//!    our state. Like how Git actually theoretically keeps track of the entire
//!    code base at every commit but the storage layer optimizations end up
//!    handling it as diffs."
//!
//! That is the whole argument for a compiled kernel under a process ontology.
//! Git's MODEL is a full tree per commit; its STORAGE is packfiles and deltas.
//! Nobody using git thinks in packfiles, and no packfile optimization is
//! allowed to change what a commit means. The model stays honest; the storage
//! gets to be clever.
//!
//! Same split here. The MODEL is "every value is folded from its whole
//! history, always, at any standpoint" — which is the honest thing and also,
//! naively, O(history) per read. The STORAGE is free to be a raster cache, a
//! packed suffix, a memoized fold, a columnar array of deltas — whatever is
//! fast — because none of it is allowed to change what a read RETURNS.
//!
//! Two things follow that are easy to miss:
//!
//!   1. The optimization has a CORRECTNESS ORACLE. The naive fold is the
//!      definition, so any clever storage can be checked against it directly
//!      (scher-state already does this: `rasterize` memoizes, `rasterize_at`
//!      stays pure and uncached as the truth-check). Optimizations you cannot
//!      check against a definition are how caches go quietly wrong.
//!   2. It only works in a COMPILED kernel. In two hand-kept twins, every
//!      storage optimization has to be written twice and agree twice — which
//!      is where the drift comes from, and why the clever version never gets
//!      built. One kernel means the storage layer can get as smart as it likes
//!      without multiplying the risk.
//!
//! So the wasm boundary is not overhead paid for tidiness. It is what makes it
//! safe to optimize the storage at all.

use crate::{is_occluded, EventRow, Society};

/// The quality marking a delta beat.
pub const Q_COUNTS: &str = "q-counts";
/// The quality marking a section placement.
pub const Q_IN_SECTION: &str = "q-in-section";

/// One delta that contributed to a value.
#[derive(Debug, Clone, PartialEq)]
pub struct Contribution {
    pub slug: String,
    pub delta: i64,
    pub by: Option<String>,
    pub at: u64,
}

/// One counted key, fully read.
#[derive(Debug, Clone, PartialEq)]
pub struct Counted {
    pub key: String,
    pub label: String,
    pub value: i64,
    pub section: Option<String>,
    pub from: Vec<Contribution>,
}

fn delta_prefix(holder: &str, key: &str) -> String {
    format!("count-{holder}-{key}-")
}

fn place_prefix(holder: &str, key: &str) -> String {
    format!("place-{holder}-{key}-")
}

/// Beats that are deltas for holder+key, oldest first.
fn delta_beats<'a>(soc: &'a Society, holder: &str, key: &str) -> Vec<&'a EventRow> {
    let p = delta_prefix(holder, key);
    let mut v: Vec<&EventRow> = soc
        .edges_onto_object(key)
        .filter(|r| r.subject.as_deref().is_some_and(|s| s.starts_with(&p)))
        .filter_map(|r| soc.get(r.subject.as_deref().unwrap_or("")))
        .collect();
    v.sort_by_key(|r| r.witnessed.unwrap_or(0));
    v
}

fn placement_beats<'a>(soc: &'a Society, holder: &str, key: &str) -> Vec<&'a EventRow> {
    let p = place_prefix(holder, key);
    let mut v: Vec<&EventRow> = soc
        .edges_onto_object(key)
        .filter(|r| r.subject.as_deref().is_some_and(|s| s.starts_with(&p)))
        .filter_map(|r| soc.get(r.subject.as_deref().unwrap_or("")))
        .collect();
    v.sort_by_key(|r| r.witnessed.unwrap_or(0));
    v
}

/// THE FOLD. A key's value at a standpoint — read, never stored.
pub fn value_of(soc: &Society, holder: &str, key: &str, as_of: Option<u64>) -> i64 {
    contributions(soc, holder, key, as_of).iter().map(|c| c.delta).sum()
}

/// The deltas that survive to this standpoint, oldest first.
pub fn contributions(
    soc: &Society,
    holder: &str,
    key: &str,
    as_of: Option<u64>,
) -> Vec<Contribution> {
    delta_beats(soc, holder, key)
        .into_iter()
        .filter(|b| as_of.is_none_or(|t| b.witnessed.unwrap_or(0) <= t))
        // an occluded delta stops counting — this IS undo, with no undo stack.
        .filter(|b| !is_occluded(soc, &b.slug, as_of))
        .map(|b| Contribution {
            slug: b.slug.clone(),
            // the delta is the beat's own content: self-describing, so the fold
            // never has to consult a second place.
            delta: b.content.parse::<i64>().unwrap_or(0),
            by: b.laid_by.clone(),
            at: b.witnessed.unwrap_or(0),
        })
        .collect()
}

/// Which section a key sits in — the newest live placement, read.
pub fn section_of(
    soc: &Society,
    holder: &str,
    key: &str,
    as_of: Option<u64>,
) -> Option<String> {
    placement_beats(soc, holder, key)
        .into_iter()
        .filter(|b| as_of.is_none_or(|t| b.witnessed.unwrap_or(0) <= t))
        .filter(|b| !is_occluded(soc, &b.slug, as_of))
        .next_back()
        .map(|b| b.content.clone())
}

/// One key, fully read.
pub fn counted_of(
    soc: &Society,
    holder: &str,
    key: &str,
    as_of: Option<u64>,
) -> Option<Counted> {
    let from = contributions(soc, holder, key, as_of);
    if delta_beats(soc, holder, key).is_empty() {
        return None;
    }
    let label = soc
        .get(key)
        .map(|b| {
            // NOTE: the Rust twin calls this field `name`; the TS twin calls
            // it `title`. Same slot, two names — a live divergence the
            // conformance corpus is what catches. Not renaming either here.
            b.name
                .clone()
                .unwrap_or_else(|| b.content.clone())
        })
        .unwrap_or_else(|| key.to_string());
    Some(Counted {
        key: key.to_string(),
        label,
        value: from.iter().map(|c| c.delta).sum(),
        section: section_of(soc, holder, key, as_of),
        from,
    })
}

/// Every key this holder counts.
pub fn keys_of(soc: &Society, holder: &str) -> Vec<String> {
    let p = format!("count-{holder}-");
    let mut keys: Vec<String> = soc
        .all()
        .filter(|r| {
            r.subject.as_deref().is_some_and(|s| s.starts_with(&p)) && r.object.is_some()
        })
        .filter_map(|r| r.object.clone())
        .collect();
    keys.sort();
    keys.dedup();
    keys
}

/// THE READ: everything this holder counts, sorted by label.
pub fn read_counted(
    soc: &Society,
    holder: &str,
    as_of: Option<u64>,
    keep_empty: bool,
) -> Vec<Counted> {
    let mut out: Vec<Counted> = keys_of(soc, holder)
        .into_iter()
        .filter_map(|k| counted_of(soc, holder, &k, as_of))
        .filter(|c| keep_empty || c.value > 0)
        .collect();
    out.sort_by(|a, b| a.label.cmp(&b.label));
    out
}

/// Sum of every value — encumbrance, total points, hand size.
pub fn total_of(soc: &Society, holder: &str, as_of: Option<u64>) -> i64 {
    read_counted(soc, holder, as_of, false).iter().map(|c| c.value).sum()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::EventRow;

    fn seeded() -> Society {
        let mut s = Society::new();
        s.lay(EventRow::node("lamp", "Lamp"));
        s.lay(EventRow::node("chart", "Moon Chart"));
        s
    }

    /// lay a delta the way the TS twin does, so both replay identically.
    fn delta(s: &mut Society, holder: &str, key: &str, n: usize, d: i64, by: &str, t: u64) {
        let slug = format!("count-{holder}-{key}-{n}");
        s.lay(
            EventRow::node(&slug, &d.to_string())
                .with_witnessed(t)
                .with_laid_by(by),
        );
        s.lay(
            EventRow::edge(&format!("{slug}~c"), "counts", &slug, key).with_witnessed(t),
        );
    }

    #[test]
    fn folds_deltas_into_a_value() {
        let mut s = seeded();
        delta(&mut s, "moth", "lamp", 0, 3, "gran", 1);
        delta(&mut s, "moth", "lamp", 1, -1, "gran", 2);
        assert_eq!(value_of(&s, "moth", "lamp", None), 2);
    }

    #[test]
    fn as_of_reads_the_past() {
        let mut s = seeded();
        delta(&mut s, "moth", "lamp", 0, 1, "gran", 1);
        delta(&mut s, "moth", "lamp", 1, 9, "gran", 5);
        assert_eq!(value_of(&s, "moth", "lamp", None), 10);
        assert_eq!(value_of(&s, "moth", "lamp", Some(1)), 1);
    }

    #[test]
    fn holders_do_not_share() {
        let mut s = seeded();
        delta(&mut s, "a", "lamp", 0, 2, "x", 1);
        delta(&mut s, "b", "lamp", 0, 7, "x", 2);
        assert_eq!(value_of(&s, "a", "lamp", None), 2);
        assert_eq!(value_of(&s, "b", "lamp", None), 7);
    }

    #[test]
    fn provenance_survives_the_fold() {
        let mut s = seeded();
        delta(&mut s, "moth", "chart", 0, 1, "gran", 1);
        delta(&mut s, "moth", "chart", 1, 1, "the-foreman", 2);
        let by: Vec<Option<String>> =
            counted_of(&s, "moth", "chart", None).unwrap().from.iter().map(|c| c.by.clone()).collect();
        assert!(by.contains(&Some("gran".into())));
        assert!(by.contains(&Some("the-foreman".into())));
    }
}

File: scher-pages/src/counted.ts
// ─────────────────────────────────────────────────────────────────────────────
// counted.ts — the general structure `stack-inventory` turned out to be an
// instance of.
//
// Hallie, 2026-07-30: "do we want a data structure that wraps an inventory and
// automatically attaches a scher-state variable to integer counts, has stack
// limits, maps to sections? … we might want genuinely scher-native data
// structures."
//
// A COUNTED is: keys whose values are FOLDED from delta-beats, optionally
// capped, optionally partitioned into sections. That is all. Its instances:
//
//   inventory      keys are items, sections are bags, cap is stack size
//   resume keys    keys are emoji, no cap, sections are resume lines
//   hp / currency  one key, a cap, no sections
//   a hand         keys are moves, cap is hand size, sections hand/deck/discard
//   appetition     keys are motivators, counted in a cone
//
// Five consumers, which is well past the two-consumer bar this package sets
// for itself.
//
// WHAT MAKES IT SCHER-NATIVE RATHER THAN A MAP WITH EXTRA STEPS
// -------------------------------------------------------------
//  1. THE VALUE IS NEVER STORED. `valueOf` folds delta-beats every read. There
//     is no field that can drift from its history, because there is no field.
//  2. MEMBERSHIP IS BETWEENNESS, NOT A FIELD. A section is a bounded interval;
//     which section a key sits in is READ by position, never written onto the
//     key. So moving between sections is an append, not a mutation — and the
//     move is in the history like everything else.
//  3. UNDO IS OCCLUSION. Occlude a delta-beat and the value re-folds. No undo
//     stack exists, because the log already is one.
//  4. AS-OF IS FREE. Every read takes a clock; the past is a read, not a
//     snapshot you kept.
//
// SEAM — the honest one: scher-state (Rust) already rasterizes typed key/value
// over a succession chain. This is the same operation in TypeScript, and the
// two could drift. The right long-term answer is probably one implementation
// compiled to WASM (scher-core-wasm exists), with this as the reference. Named
// now rather than discovered later.
// ─────────────────────────────────────────────────────────────────────────────

import { Society, isOccluded } from "scher/society";

/** A key in a Counted. Opaque — never parsed for meaning (no string-matching
 *  on slugs; that is the discipline scher's CLAUDE.md names explicitly). */
export type Key = string;

// Constants and slug schemes are GENERATED from the Rust kernel — the parts
// that drift silently (a renamed quality or a changed scheme does not throw,
// it just makes the fold read nothing, which looks like "you have zero").
// Regenerate: node scripts/generate-from-rust.mjs · CI check: --check
export { Q_COUNTS, Q_IN_SECTION } from "./counted.generated.js";
import { deltaPrefix, placePrefix, Q_COUNTS as Q_COUNTS_V,
         Q_IN_SECTION as Q_IN_SECTION_V } from "./counted.generated.js";

export interface CountedSpec {
  /** namespace: whose counted is this. Two holders never share by accident. */
  holder: string;
  /** per-key cap. A number applies to every key; a function lets one key differ.
   *  Over-cap writes are REFUSED, never silently clamped — a clamped write is a
   *  lie the history cannot show. */
  cap?: number | ((key: Key) => number | undefined);
  /** the named sections, in display order. Omit for an unsectioned counted. */
  sections?: string[];
  /** default section for a newly-counted key. */
  defaultSection?: string;
}

export interface CountedRead {
  /** read as of this clock; omit for now. */
  asOf?: number;
  /** keep keys that folded to zero (a WoW bag keeps the empty slot). */
  keepEmpty?: boolean;
}

/** One counted key, folded. */
export interface Counted {
  key: Key;
  label: string;
  value: number;
  /** which section it is in, READ by placement — never a stored field. */
  section: string | null;
  cap?: number;
  /** the deltas that produced the value, newest first. */
  from: Array<{ slug: string; delta: number; by: string | null; at: number }>;
}

const capFor = (spec: CountedSpec, key: Key): number | undefined =>
  typeof spec.cap === "function" ? spec.cap(key) : spec.cap;

// ── WRITES: every one of them is an append ──────────────────────────────────

/** Add `delta` to a key (negative to subtract).
 *  Returns the beat slug, or null if refused. */
export function count(soc: Society, spec: CountedSpec, key: Key,
                      delta: number, by?: string): string | null {
  if (!Number.isFinite(delta) || delta === 0) return null;

  const now = valueOf(soc, spec, key);
  if (now + delta < 0) return null;                       // never below zero
  const cap = capFor(spec, key);
  if (cap !== undefined && now + delta > cap) return null; // refuse, not clamp

  const n = deltaBeats(soc, spec.holder, key).length;
  const slug = `${deltaPrefix(spec.holder, key)}${n}`;
  soc.lay({
    // the delta is the beat's own content: self-describing, so the fold never
    // has to consult a second place.
    slug, content: String(delta), title: null,
    subject: null, object: null, laid_by: by ?? null,
  });
  soc.layP(`${slug}~c`, `${delta > 0 ? "+" : ""}${delta} ${key}`, slug, key, Q_COUNTS_V);

  // first sighting of a key lands it in the default section, as an append.
  if (spec.defaultSection && sectionOf(soc, spec, key) === null)
    place(soc, spec, key, spec.defaultSection, by);

  return slug;
}

/** Put a key in a section. An APPEND — the newest placement wins the read, and
 *  every previous placement stays visible in history. */
export function place(soc: Society, spec: CountedSpec, key: Key,
                      section: string, by?: string): string | null {
  if (spec.sections && !spec.sections.includes(section)) return null;
  const n = placementBeats(soc, spec.holder, key).length;
  const slug = `${placePrefix(spec.holder, key)}${n}`;
  soc.lay({ slug, content: section, title: null, subject: null, object: null, laid_by: by ?? null });
  soc.layP(`${slug}~p`, `${key} → ${section}`, slug, key, Q_IN_SECTION_V);
  return slug;
}

/** Move `n` of a key from one section to another. Split-friendly: moving part
 *  of a stack spends here and acquires there, so both halves keep provenance. */
export function moveBetween(soc: Society, spec: CountedSpec, key: Key,
                            to: string, by?: string): string | null {
  return place(soc, spec, key, to, by);
}

/** Undo any counted write: occlude it and the value re-folds. */
export function undo(soc: Society, slug: string, lesson: string, by?: string): string | null {
  if (!soc.get(slug) || !lesson.trim()) return null;
  const occ = `occ-${slug}`;
  if (soc.get(occ)) return null;               // already undone; inert
  soc.lay({ slug: occ, content: lesson, subject: null, object: null, laid_by: by ?? null });
  soc.layP(`${occ}~o`, lesson, occ, slug, "q-occludes");
  return occ;
}

// ── READS: nothing below is stored ──────────────────────────────────────────

function deltaBeats(soc: Society, holder: string, key: Key) {
  const p = deltaPrefix(holder, key);
  return soc.all()
    .filter((r) => r.subject !== null && r.object === key && (r.subject as string).startsWith(p))
    .sort((a, b) => (a.witnessed ?? 0) - (b.witnessed ?? 0));
}

function placementBeats(soc: Society, holder: string, key: Key) {
  const p = placePrefix(holder, key);
  return soc.all()
    .filter((r) => r.subject !== null && r.object === key && (r.subject as string).startsWith(p))
    .sort((a, b) => (a.witnessed ?? 0) - (b.witnessed ?? 0));
}

/** THE FOLD. A key's value, read — never stored. */
export function valueOf(soc: Society, spec: CountedSpec, key: Key, opts: CountedRead = {}): number {
  let v = 0;
  for (const e of deltaBeats(soc, spec.holder, key)) {
    const beat = soc.get(e.subject as string);
    if (!beat) continue;
    if (opts.asOf !== undefined && (beat.witnessed ?? 0) > opts.asOf) continue;
    // asOf MUST be passed: occlusion is standpoint-relative, and reading it
    // at "now" while reading deltas "as of then" makes the past change every
    // time the present does. (Caught by conformance/counted.json, not review.)
    if (isOccluded(soc, beat.slug, opts.asOf) || isOccluded(soc, e.slug, opts.asOf)) continue;
    v += Number(beat.content) || 0;
  }
  return v;
}

/** Which section a key sits in — READ from the newest live placement. */
export function sectionOf(soc: Society, spec: CountedSpec, key: Key,
                          opts: CountedRead = {}): string | null {
  const live = placementBeats(soc, spec.holder, key)
    .map((e) => soc.get(e.subject as string))
    .filter((b): b is NonNullable<typeof b> => !!b)
    .filter((b) => opts.asOf === undefined || (b.witnessed ?? 0) <= opts.asOf)
    .filter((b) => !isOccluded(soc, b.slug, opts.asOf));
  return live.length ? live[live.length - 1].content : null;
}

/** One key, fully read. */
export function countedOf(soc: Society, spec: CountedSpec, key: Key,
                          opts: CountedRead = {}): Counted | null {
  const beats = deltaBeats(soc, spec.holder, key);
  if (!beats.length) return null;
  const from: Counted["from"] = [];
  for (const e of beats) {
    const b = soc.get(e.subject as string);
    if (!b) continue;
    const at = b.witnessed ?? 0;
    if (opts.asOf !== undefined && at > opts.asOf) continue;
    if (isOccluded(soc, b.slug, opts.asOf) || isOccluded(soc, e.slug, opts.asOf)) continue;
    from.push({ slug: b.slug, delta: Number(b.content) || 0, by: b.laid_by ?? null, at });
  }
  from.reverse();
  return {
    key,
    label: soc.get(key)?.title ?? soc.get(key)?.content ?? key,
    value: from.reduce((n, f) => n + f.delta, 0),
    section: sectionOf(soc, spec, key, opts),
    cap: capFor(spec, key),
    from,
  };
}

/** Every key this counted holds. */
export function keysOf(soc: Society, spec: CountedSpec): Key[] {
  const p = `count-${spec.holder}-`;
  const out = new Set<Key>();
  for (const r of soc.all())
    if (r.subject && r.object && (r.subject as string).startsWith(p)) out.add(r.object as string);
  return [...out];
}

/** THE READ: everything, optionally grouped by section. */
export function readCounted(soc: Society, spec: CountedSpec, opts: CountedRead = {}): Counted[] {
  return keysOf(soc, spec)
    .map((k) => countedOf(soc, spec, k, opts))
    .filter((c): c is Counted => !!c)
    .filter((c) => opts.keepEmpty || c.value > 0)
    .sort((a, b) => a.label.localeCompare(b.label));
}

/** Grouped by section, in the spec's declared section order. */
export function bySection(soc: Society, spec: CountedSpec,
                          opts: CountedRead = {}): Array<{ section: string; items: Counted[] }> {
  const all = readCounted(soc, spec, opts);
  const names = spec.sections ?? [...new Set(all.map((c) => c.section ?? ""))];
  return names.map((section) => ({
    section,
    items: all.filter((c) => (c.section ?? "") === section),
  }));
}

/** Sum of every value — encumbrance, total points, hand size. */
export function totalOf(soc: Society, spec: CountedSpec, opts: CountedRead = {}): number {
  return readCounted(soc, spec, opts).reduce((n, c) => n + c.value, 0);
}
